test/unit/math/prim/mat/fun/add_diag_test.cpp

```diff
@@ -0,0 +1,156 @@
+}
+
+//  add vector to diag
+
+TEST(MathPrimMat, double_mat_double_vec_add_diag) {
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> mat(2, 3);
+  mat << 1, 1, 1, 1, 1, 1;
+
+  Eigen::Matrix<double, 1, -1> to_add(2);
+  to_add << 0, 1;
+
+  Eigen::MatrixXd out_mat;
+  EXPECT_NO_THROW(out_mat = stan::math::add_diag(mat, to_add));
+  for (int i = 0; i < 2; ++i)
+    EXPECT_FLOAT_EQ(1 + to_add[i], out_mat(i, i))
+        << "index: ( " << i << ", " << i << ")";
+}
+
+TEST(MathPrimMat, double_mat_double_rvec_add_diag) {
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> mat(2, 3);
+  mat << 1, 1, 1, 1, 1, 1;
+
+  Eigen::Matrix<double, -1, 1> to_add(2);
+  to_add << 0, 1;
+
+  Eigen::MatrixXd out_mat;
+  EXPECT_NO_THROW(out_mat = stan::math::add_diag(mat, to_add));
+  for (int i = 0; i < 2; ++i)
+    EXPECT_FLOAT_EQ(1 + to_add[i], out_mat(i, i))
+        << "index: ( " << i << ", " << i << ")";
+}
+
+TEST(MathPrimMat, var_mat_double_vec_add_diag) {
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> mat(2, 3);
+  mat << 1, 1, 1, 1, 1, 1;
+
+  Eigen::Matrix<double, 1, -1> to_add(2);
+  to_add << 0, 1;
+
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> out_mat;
+  EXPECT_NO_THROW(out_mat = stan::math::add_diag(mat, to_add));
+  for (int i = 0; i < 2; ++i)
+    EXPECT_FLOAT_EQ(1 + to_add[i], value_of(out_mat(i, i)))
+        << "index: ( " << i << ", " << i << ")";
+}
+
+TEST(MathPrimMat, var_mat_double_rvec_add_diag) {
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> mat(2, 3);
+  mat << 1, 1, 1, 1, 1, 1;
+
+  Eigen::Matrix<double, -1, 1> to_add(2);
+  to_add << 0, 1;
+
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> out_mat;
+  EXPECT_NO_THROW(out_mat = stan::math::add_diag(mat, to_add));
+  for (int i = 0; i < 2; ++i)
+    EXPECT_FLOAT_EQ(1 + to_add[i], value_of(out_mat(i, i)))
+        << "index: ( " << i << ", " << i << ")";
+}
+
+TEST(MathPrimMat, double_mat_var_vec_add_diag) {
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> mat(2, 3);
+  mat << 1, 1, 1, 1, 1, 1;
+
+  Eigen::Matrix<stan::math::var, 1, -1> to_add(2);
+  to_add << 0, 1;
+
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> out_mat;
+  EXPECT_NO_THROW(out_mat = stan::math::add_diag(mat, to_add));
+  for (int i = 0; i < 2; ++i)
+    EXPECT_FLOAT_EQ(1 + value_of(to_add[i]), value_of(out_mat(i, i)))
+        << "index: ( " << i << ", " << i << ")";
+}
+
+TEST(MathPrimMat, double_mat_var_rvec_add_diag) {
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> mat(2, 3);
+  mat << 1, 1, 1, 1, 1, 1;
+
+  Eigen::Matrix<stan::math::var, -1, 1> to_add(2);
+  to_add << 0, 1;
+
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> out_mat;
+  EXPECT_NO_THROW(out_mat = stan::math::add_diag(mat, to_add));
+  for (int i = 0; i < 2; ++i)
+    EXPECT_FLOAT_EQ(1 + value_of(to_add[i]), value_of(out_mat(i, i)))
+        << "index: ( " << i << ", " << i << ")";
+}
```

Comment:
@syclik --- do you want tests for derivatives if everything is templated in prim?  All the derivatives should work if the primitive definitions only use Stan differentiable functions.  If they don't, it's a sign of another bug in the API somewhere.  So I'd be OK without `rev` and `fwd` tests here.  

We do need the instantiaton tests.